### ISSUE TITLE
Add support for additional log drivers

### ISSUE THREAD

**Issue by haircommander** [containers/conmon-rs#1126]
Currently CRI format is the only one supported. At minimum long term, json-file and journald should also be added. https://github.com/containers/conmon-rs/issues/482 tracks adding splunk as a log driver as well

**Comment by validsudoku** [comment]
@haircommander @saschagrunert are there resources (links, code files etc) one can look at to understand this issue much better?

**Comment by saschagrunert** [comment]
Hey, most of the container log driver abstraction lives in

https://github.com/containers/conmon-rs/blob/634fd10acdcd350e784e1d068a783b773cdee4a9/conmon-rs/server/src/container_log.rs#L12-L19

Means we would require to adapt this part, the cap n proto interface

https://github.com/containers/conmon-rs/blob/634fd10acdcd350e784e1d068a783b773cdee4a9/conmon-rs/common/proto/conmon.capnp#L42-L56

as well as the client including testing

https://github.com/containers/conmon-rs/blob/634fd10acdcd350e784e1d068a783b773cdee4a9/pkg/client/client.go#L632-L652

**Comment by TimePrinciple** [comment]
Hi, I'm interested in this idea.

I've just `cargo test`ed  `conmon-rs`, and I didn't see any UTs in file `conmon-rs/conmon-rs/server/src/container_log.rs`, thus not quite sure what kind of output is desired.

I wonder the works to be done is to reformat the output `CriLogger` to JSON and splunk, or to create different `ContainerRuntimeInterface` for both.

**Comment by jellllly420** [comment]
Hi, I'm interested in this project.

I'm currently not so familiar with the details in conmon-rs. However, from the structure of the code, I guess the works to be done in this project is to create different loggers like the `cri_logger` and add them to the tuple enum `LogDriver` and at last modify the related cap n proto and go parts. If there is anything wrong, please let me know. :)

Hope to join and finish this project.

**Comment by aashi-ihsaa** [comment]
@haircommander I have experience in rust and containerization and this issue seems like a good way to go about that. I would like to take this up under LFX term 3. Can you just a bit elaborate on what kind of output is desired

**Comment by octonawish-akcodes** [comment]
I am interested in this issue under LFX sep term

**Comment by octonawish-akcodes** [comment]
@haircommander Can you give some guidance on what to look up for, so that I could've a better understanding of this issue.

**Comment by haircommander** [comment]
the resulting output will be equivalent log driver behavior to Docker's https://docs.docker.com/config/containers/logging/configure/  for some or all of the aforementioned drivers

**Comment by haircommander** [comment]
what level is your understanding of the ecosystem @octonawish-akcodes ?

**Comment by royalpinto007** [comment]
Hey @haircommander,

I was going through the issue and found that I could work on this problem. I have previously dealt with a similar matter in one of my past experiences (specifically, adding language support to a VSCode Extension). This background makes me a suitable candidate for this task, don't you think?

By the way, I'm applying as an LFX mentee for the upcoming fall term, and I am very interested in contributing. 
Could you please provide any relevant documentation if available? If not, I'll take the initiative to explore further and get started. I hope that's okay with you :)

**Comment by prash240303** [comment]
hello @haircommander  @saschagrunert  , i tried to understand the functions for the criologger and read the docker docs ,and tried to write the same for the json logger (also i couldnt get the crio setup locally to test this out with conmon , so , its just skimming on the codebase i read and tried to understand) ihope things work this way and for journald logger too , also i couldnt understand the client side testing

```

#[derive(Debug, Default)]
pub struct ContainerLog {
    drivers: Vec<LogDriver>,
}

#[derive(Debug)]
enum LogDriver {
    ContainerRuntimeInterface(CriLogger),
    JsonLogger(JsonLogger), // New JSON logger variant
}

// New JSON logger implementation
#[derive(Debug)]
struct JsonLogger {
    type @0 :Type; 
    path @1 :Text; 
    max_size @2 :UInt64; 
    }

```


Tho i am not sure with the implementaion of json logger (so left the functions empty)
```

impl JsonLogger {
    fn new(path: String, max_size: u64) -> Result<Self> { 
        Ok(Self {
            path,
            max_size,
            //function here
        })
    }

    fn init(&mut self) -> Result<()> {
     //json logger initalization
        Ok(())
    }

    fn reopen(&mut self) -> Result<()> {
      //reopen the resources for json logging 
        Ok(())
    }

    fn write<T>(&mut self, _pipe: Pipe, _bytes: T) -> Result<()>
    where
        T: AsyncBufRead + Unpin + Copy,
    {
      //writing json logs
        Ok(())
    }
}

container log implementation impl ContainerLog {
    // ... (existing methods)

    pub fn from(reader: Reader<Owned>) -> Result<SharedContainerLog> {
        let drivers = reader
            .iter()
            .flat_map(|x| -> Result<_> {
                Ok(match x.get_type()? {
                    Type::ContainerRuntimeInterface => {
                        LogDriver::ContainerRuntimeInterface(CriLogger::new(
                            x.get_path()?,
                            if x.get_max_size() > 0 {
                                Some(x.get_max_size() as usize)
                            } else {
                                None
                            },
                        )?)
                    }
                    Type::JsonLogger => {
                        LogDriver::JsonLogger(JsonLogger::new(x.get_path()?)?)
                    }
                })
            })
            .collect();
        Ok(Arc::new(RwLock::new(Self { drivers })))
    }
    
    // ... (existing methods)

    // New method to initialize JSON loggers
    pub async fn init_json_loggers(&mut self) -> Result<()> {
        join_all(
            self.drivers
                .iter_mut()
                .filter_map(|x| {
                    if let LogDriver::JsonLogger(json_logger) = x {
                        Some(json_logger.init())
                    } else {
                        None
                    }
                })
                .collect::<Vec<_>>(),
        )
        .await
        .into_iter()
        .collect::<Result<Vec<_>>>()?;
        Ok(())
    }

    // New method to reopen JSON loggers
    pub async fn reopen_json_loggers(&mut self) -> Result<()> {
        join_all(
            self.drivers
                .iter_mut()
                .filter_map(|x| {
                    if let LogDriver::JsonLogger(json_logger) = x {
                        Some(json_logger.reopen())
                    } else {
                        None
                    }
                })
                .collect::<Vec<_>>(),
        )
        .await
        .into_iter()
        .collect::<Result<Vec<_>>>()?;
        Ok(())
    }

    // New method to write JSON logs
    pub async fn write_json_logs<T>(&mut self, pipe: Pipe, bytes: T) -> Result<()>
    where
        T: AsyncBufRead + Unpin + Copy,
    {
        join_all(
            self.drivers
                .iter_mut()
                .filter_map(|x| {
                    if let LogDriver::JsonLogger(json_logger) = x {
                        Some(json_logger.write(pipe, bytes))
                    } else {
                        None
                    }
                })
                .collect::<Vec<_>>(),
        )
        .await
        .into_iter()
        .collect::<Result<Vec<_>>>()?;
        Ok(())
    }
}

```

**Comment by ItsOrganic** [comment]
Hey @haircommander LFX Fall 23 applicant here, I would like to contribute to this project as I have working knowledge of Rust and containerization. We just need to create new separate modules for each and support logs formatting say for JSON we need to format logs data to JSON including metadata and timestamps similarly for Journald and splunk.

**Comment by d4v1d03** [comment]
Hey there @haircommander, 
I've filled the Fall application as of now and would like you to know my proficiency in RUST and container runtime technologies such as K8s,CRI-O and containerd. I also have a solid understanding of runC and its role as a default runtime for Docker and K8s, enabling the execution of containers according to OCI specifications. I was exploring the codebase of conmon-rs and would like to work over it. I've started the work for the JSON driver and would try my best to make a PR in the upcoming week.
